File: 121_best_time_to_buy_and_sell_stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int max=0,max_temp=0;
        vector<int> diff;
        for (int i=1;i<prices.size();i++) 
            diff.push_back(prices[i]-prices[i-1]);
        for (int i=0;i<diff.size();i++) {
            max_temp=max_temp+diff[i];
            if (diff[i]>=0) max=max>max_temp?max:max_temp;
            else if (max_temp<=0) max_temp=0;
        }
        return max;
    }
};

File: algorithms/python3/121_best_time_to_buy_and_sell_stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        max1=prices.copy()
        max2=prices.copy()
        for i in range(1,len(prices)):
            if max1[i]>max1[i-1]:
                max1[i]=max1[i-1]
            if max2[len(prices)-i-1]<max2[len(prices)-i]:
                max2[len(prices)-i-1]=max2[len(prices)-i]
        return max([a-b for a,b in zip(max2,max1)])
    def maxProfit2(self, prices: List[int]) -> int:
        maxprofit=0
        minprice=float('inf')
        for i in range(len(prices)):
            if prices[i]<minprice:
                minprice=prices[i]
            elif prices[i]-minprice>maxprofit:
                maxprofit=prices[i]-minprice
        return maxprofit
a=Solution()

print(a.maxProfit2([7,1,5,3,6,4]))

